GH-8034 Fixing infinite scroll e2e test

hackergil opened the pull request:
- Getting number of results from xhr call instead of breadcrumb
- A bit of refactoring (using constants, moving logic to the test itself, structuring test like the others)

Fixes #8034

hackergil committed:
GH-8034 Fixing infinite scroll e2e test
- Getting number of results from xhr call instead of breadcrumb
- A bit of refactoring (using constants, moving logic to the test itself, structuring test like the others)

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/infinite-scroll.e2e-spec.ts:
```diff
@@ -1,50 +1,84 @@
     });
+  });
+
+  it('Should not display Infinite scroll', () => {
```
small s `should` like the others

hackergil committed:
Merge branch 'develop' into feature/GH-8034

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/infinite-scroll.e2e-spec.ts:
```diff
@@ -1,50 +1,84 @@
+
+      verifySortingResetsList();
+
+      verifyFilterResetsList();
```
Running on the test runner or headless x amount of times fails due to the general endpoint. It wasn't failing before until the change of breadcrumb rendering performance, but I think having it refactored like this triggered it maybe. 

I believe you would need to target the specific url for sorting / filterring / gridreset - search endpoint

If it runs only have 1 of them, the test passes sometimes, but if we have two or more of those assertions to that route (verifySortingResetsList/verifyFilterResetsList/verifyGridResetsList) , then it says the screenshot below, where it fails sometimes on the verifySortingResetsList function. I believe the fix for specifying specifics will help!

- tested on 2005 server, develop, and dev-6 in both the testrunner and headless version

<img width="443" alt="Screen Shot 2020-07-08 at 5 22 46 PM" src="https://user-images.githubusercontent.com/29099466/86971736-ac894880-c13f-11ea-833a-87aaa9a9c106.png">

bgambocjaviniar replied:
hmm tried it again, and it passed, I would say it's flaky 
<img width="404" alt="Screen Shot 2020-07-08 at 5 38 00 PM" src="https://user-images.githubusercontent.com/29099466/86973080-f70bc480-c141-11ea-8477-7590f533e713.png">
<img width="398" alt="Screen Shot 2020-07-08 at 5 38 10 PM" src="https://user-images.githubusercontent.com/29099466/86973153-1c98ce00-c142-11ea-868b-7c1059c3cd21.png">

bgambocjaviniar replied:
jenkins build #jenkins-394 https://jkmaster.test.c3po.b2c.ydev.hybris.com/job/spartacus-regression-tests-for-branch/394/consoleFull shows a new error that I never seen running it

jenkins build #jenkins-395-396 displays the same result I found above https://jkmaster.test.c3po.b2c.ydev.hybris.com/job/spartacus-regression-tests-for-branch/395/consoleFull
https://jkmaster.test.c3po.b2c.ydev.hybris.com/job/spartacus-regression-tests-for-branch/396/consoleFull

hackergil replied:
Great findings Brian, thank you. Addressed now.

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/infinite-scroll.e2e-spec.ts:
```diff
@@ -1,50 +1,84 @@
-      scrollConfig(true, 0, true);
-      verifyProductListLoaded();
-    });
+  it("should enable Infinite scroll and NOT display 'Show more' button", () => {
```
I don't mind having the test like this, but I guess it will be a topic for tomorrow when explaining the best practices, but wouldn't it be best to isolate what is it doing like in it's own `its` for assertions which enables us to find where it failed easier or no?

bgambocjaviniar replied:
other than that, I believe it's fine after talking about the recommendations if we need to update this test to get # from response then check the breadcrumb if it corresponds

bgambocjaviniar replied:
^ was thinking about the first point, but scratch what I said. Having it produce the end result from the description is all we want

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/infinite-scroll.e2e-spec.ts:
```diff
@@ -1,50 +1,84 @@
     });
+  });
+
+  it("Should enable infinite scroll and display 'Show more' button", () => {
```
same as below*

hackergil committed:
Merge branch 'develop' into feature/GH-8034

hackergil committed:
[GH-8034] Addressing code review changes
- Creating more specific routes for default, sort, brand queries
- Removing some dependencies from product search helper

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/infinite-scroll.e2e-spec.ts:
```diff
@@ -78,7 +76,7 @@ describe('Infinite scroll', () => {
     configScroll(false, 0, false);
```
line 77 'should'

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/helpers/infinite-scroll.ts:
```diff
@@ -104,35 +112,37 @@ export function scrollToFooter(
 }
 
 export function verifyFilterResetsList() {
+  cy.route(
```
I think this route should be moved to the createGridQuery() function above in line 39
And call that function at the toplevel of the `it`, so that there's more flexibility in the future and for re-usability since if it's kept here, what if one day we want to call verifyGridResetList

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/helpers/infinite-scroll.ts:
```diff
@@ -104,35 +112,37 @@ export function scrollToFooter(
 }
 
 export function verifySortingResetsList() {
+  cy.route(
```
not necessary like the comment above/below, but maybe we can pull this out to put in a function and call it as well in the top level of the `it`

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/helpers/infinite-scroll.ts:
```diff
@@ -30,6 +29,15 @@ export function configScroll(
+  ).as(defaultQueryName);
+}
+
+export function createGridQuery() {}
```
be populated by below/above comment from  line 129

hackergil committed:
[GH-8034] Moving queries for better maintenance

hackergil committed:
prettier

hackergil committed:
more prettier

bgambocjaviniar commented on projects/storefrontapp-e2e-cypress/cypress/helpers/infinite-scroll.ts:
```diff
@@ -112,10 +112,7 @@ export function scrollToFooter(
 }
```
createGridQuery 
line 39 need to delete since it's empty

bgambocjaviniar committed:
removed unused function

bgambocjaviniar committed:
Merge branch 'develop' into feature/GH-8034
